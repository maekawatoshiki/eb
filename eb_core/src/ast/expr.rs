use crate::lexer::{location::Location, token::PunctKind};

#[derive(Debug, Clone, PartialEq)]
pub struct Node {
    kind: Kind,
    loc: Location,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Kind {
    Int(i64),
    Ident(String),
    BinOp(BinOpKind, Box<Node>, Box<Node>),
}

#[derive(Debug, Clone, PartialEq)]
pub enum BinOpKind {
    Add,
    Sub,
    Mul,
    Div,
    Eq,
}

impl Node {
    pub fn new(kind: Kind, loc: Location) -> Self {
        Self { kind, loc }
    }
}

impl From<Node> for super::Node {
    fn from(node: Node) -> Self {
        super::Node::Expr(node)
    }
}

impl From<PunctKind> for BinOpKind {
    fn from(p: PunctKind) -> Self {
        match p {
            PunctKind::Plus => Self::Add,
            PunctKind::Minus => Self::Sub,
            PunctKind::Eq => Self::Eq,
            _ => panic!(),
        }
    }
}
