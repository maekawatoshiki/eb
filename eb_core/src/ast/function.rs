use crate::lexer::location::Location;

#[derive(Debug, Clone, PartialEq)]
pub struct Node {
    name: String,
    loc: Location,
}

impl Node {
    pub fn new(name: String, loc: Location) -> Self {
        Self { name, loc }
    }
}

impl From<Node> for super::Node {
    fn from(node: Node) -> Self {
        super::Node::Function(node)
    }
}
