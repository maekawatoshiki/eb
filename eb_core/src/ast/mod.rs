pub mod expr;
pub mod function;

#[derive(Debug, PartialEq, Clone)]
pub enum Node {
    Function(function::Node),
    Expr(expr::Node),
}
