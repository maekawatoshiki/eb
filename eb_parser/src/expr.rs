use super::{function, Context, Error};
use crate::{
    ast::expr,
    lexer::token::{DelimKind, PunctKind, TokenKind},
};
use anyhow::Result;

pub fn parse(ctx: &mut Context) -> Result<expr::Node> {
    parse_binop_eq_ne(ctx)
}

fn parse_binop_eq_ne(ctx: &mut Context) -> Result<expr::Node> {
    let mut lhs = parse_binop_add_sub(ctx)?;
    loop {
        let loc = ctx.cur_loc();

        let eq = ctx.skip_punct(PunctKind::Eq);
        let neq = ctx.skip_punct(PunctKind::Neq);

        if !eq && !neq {
            break;
        }

        let loc = loc?;
        let rhs = parse_binop_add_sub(ctx)?;

        lhs = expr::Node::new(
            expr::Kind::BinOp(
                if eq {
                    expr::BinOpKind::Eq
                } else {
                    expr::BinOpKind::Neq
                },
                Box::new(lhs.clone()),
                Box::new(rhs),
            ),
            loc,
        );
    }
    Ok(lhs)
}

fn parse_binop_add_sub(ctx: &mut Context) -> Result<expr::Node> {
    let mut lhs = parse_binop_mul_div(ctx)?;
    loop {
        let loc = ctx.cur_loc();

        let plus = ctx.skip_punct(PunctKind::Plus);
        let minus = ctx.skip_punct(PunctKind::Minus);

        if !plus && !minus {
            break;
        }

        let loc = loc?;
        let rhs = parse_binop_mul_div(ctx)?;

        lhs = expr::Node::new(
            expr::Kind::BinOp(
                if plus {
                    expr::BinOpKind::Add
                } else {
                    expr::BinOpKind::Sub
                },
                Box::new(lhs.clone()),
                Box::new(rhs),
            ),
            loc,
        );
    }
    Ok(lhs)
}

fn parse_binop_mul_div(ctx: &mut Context) -> Result<expr::Node> {
    let mut lhs = parse_postfix(ctx)?;
    loop {
        let loc = ctx.cur_loc();

        let star = ctx.skip_punct(PunctKind::Star);
        let slash = ctx.skip_punct(PunctKind::Slash);

        if !star && !slash {
            break;
        }

        let loc = loc?;
        let rhs = parse_postfix(ctx)?;

        lhs = expr::Node::new(
            expr::Kind::BinOp(
                if star {
                    expr::BinOpKind::Mul
                } else {
                    expr::BinOpKind::Div
                },
                Box::new(lhs.clone()),
                Box::new(rhs),
            ),
            loc,
        );
    }
    Ok(lhs)
}

fn parse_postfix(ctx: &mut Context) -> Result<expr::Node> {
    let base = parse_primary(ctx)?;
    let peek = match ctx.peek() {
        Some(peek) => peek,
        None => return Ok(base),
    };
    let loc = *peek.loc();
    match peek.kind() {
        // Call
        TokenKind::OpenDelim(DelimKind::Paren) => {
            assert!(ctx.next().is_some());
            Ok(expr::Node::new(
                expr::Kind::Call(Box::new(base), parse_call_args(ctx)?),
                loc,
            ))
        }
        _ => Ok(base),
    }
}

fn parse_call_args(ctx: &mut Context) -> Result<Vec<expr::Node>> {
    if ctx.skip_close_delim(DelimKind::Paren) {
        return Ok(vec![]);
    }

    let mut args = vec![];

    loop {
        let arg = parse(ctx)?;
        args.push(arg);

        if ctx.skip_punct(PunctKind::Comma) {
            continue;
        }

        ctx.expect_close_delim(DelimKind::Paren)?;

        break;
    }

    Ok(args)
}

fn parse_primary(ctx: &mut Context) -> Result<expr::Node> {
    let peek = ctx.peek().ok_or(Error::EOF)?;
    let loc = *peek.loc();
    match peek.kind() {
        TokenKind::Int(int) => {
            let int = int.parse().unwrap();
            ctx.next().unwrap();
            Ok(expr::Node::new(expr::Kind::Int(int), loc))
        }
        TokenKind::Ident(ident) if ident == &"func" => Ok(expr::Node::new(
            expr::Kind::Function(Box::new(function::parse(ctx)?)),
            loc,
        )),
        TokenKind::Ident(ident) if ident == &"if" => Ok(expr::Node::new(parse_if(ctx)?, loc)),
        TokenKind::Ident(ident) => {
            let ident = ident.to_string();
            ctx.next().unwrap();
            Ok(expr::Node::new(expr::Kind::Ident(ident), loc))
        }
        _ => Err(Error::ExpectedAny(loc, "integer value or identifier").into()),
    }
}

fn parse_if(ctx: &mut Context) -> Result<expr::Kind> {
    ctx.expect_keyword("if")?;
    let cond = parse(ctx)?;
    ctx.expect_punct(PunctKind::Colon)?;
    let loc = ctx.cur_loc()?;
    let then_expr = expr::Node::new(expr::Kind::Exprs(parse_body(ctx)?), loc);
    let else_expr;
    if ctx.skip_keyword("else") {
        ctx.expect_punct(PunctKind::Colon)?;
        let loc = ctx.cur_loc()?;
        else_expr = Some(Box::new(expr::Node::new(
            expr::Kind::Exprs(parse_body(ctx)?),
            loc,
        )));
    } else {
        else_expr = None;
    }
    Ok(expr::Kind::If(
        Box::new(cond),
        Box::new(then_expr),
        else_expr,
    ))
}

pub fn parse_body(ctx: &mut Context) -> Result<Vec<expr::Node>> {
    if ctx.skip_punct(PunctKind::DoubleSemicolon) {
        return Ok(vec![]);
    }

    let mut body = vec![];

    loop {
        body.push(parse(ctx)?);

        if ctx.skip_punct(PunctKind::Semicolon) {
            continue;
        }

        if ctx.skip_punct(PunctKind::DoubleSemicolon) {
            return Ok(body);
        }
    }
}

#[cfg(test)]
mod test {
    extern crate insta;
    use super::*;
    use crate::lexer::{source::Source, tokenize};

    fn parse_str(s: &str) -> expr::Node {
        let source = Source::String(s.to_string());
        let mut ctx = Context::new(tokenize(&source));
        parse(&mut ctx).expect("fail to parse")
    }

    #[test]
    fn parse1() {
        insta::assert_debug_snapshot!(parse_str(r#"x"#));
    }

    #[test]
    fn parse2() {
        insta::assert_debug_snapshot!(parse_str(r#"x +x"#));
    }

    #[test]
    fn parse3() {
        insta::assert_debug_snapshot!(parse_str(r#"123 + x"#));
    }

    #[test]
    fn parse4() {
        insta::assert_debug_snapshot!(parse_str(r#"1 * 2 + 3"#));
    }

    #[test]
    fn parse5() {
        insta::assert_debug_snapshot!(parse_str(r#"f()"#));
    }

    #[test]
    fn parse6() {
        insta::assert_debug_snapshot!(parse_str(r#"f(1, x)"#));
    }

    #[test]
    fn parse7() {
        insta::assert_debug_snapshot!(parse_str(r#"x == x"#));
    }

    #[test]
    fn parse8() {
        insta::assert_debug_snapshot!(parse_str(r#"x != x"#));
    }

    #[test]
    fn parse9() {
        insta::assert_debug_snapshot!(parse_str(
            r#"if x == 1:
                123;;"#
        ));
    }

    #[test]
    fn parse10() {
        insta::assert_debug_snapshot!(parse_str(
            r#"if x == 1:
                 x + 1 ; 
                 x ;;
               else:
                 42;;"#
        ));
    }
}
