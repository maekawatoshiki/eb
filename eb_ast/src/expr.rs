use super::function;
use lexer::location::Location;

#[derive(Debug, Clone, PartialEq)]
pub struct Node {
    kind: Kind,
    loc: Location,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Kind {
    Int(i64),
    Ident(String),
    Function(Box<function::Node>),
    BinOp(BinOpKind, Box<Node>, Box<Node>),
    Call(Box<Node>, Vec<Node>),
    If(Box<Node>, Box<Node>, Option<Box<Node>>),
    Exprs(Vec<Node>),
}

#[derive(Debug, Clone, PartialEq)]
pub enum BinOpKind {
    Add,
    Sub,
    Mul,
    Div,
    Eq,
    Neq,
}

impl Node {
    pub fn new(kind: Kind, loc: Location) -> Self {
        Self { kind, loc }
    }
}
