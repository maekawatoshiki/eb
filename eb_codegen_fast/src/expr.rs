use super::function::{self, Context};
use anyhow::Result;
use ast::expr as ast_expr;
use vm::inst::Inst;

pub fn visit(ctx: &mut Context, expr: &ast_expr::Node) -> Result<()> {
    match expr.kind() {
        ast_expr::Kind::Int(i) => {
            ctx.push(Inst::PushInt(*i as i32)); // TODO
        }
        ast_expr::Kind::Ident(ident) => {
            ctx.push(Inst::PushStr(ident.to_owned()));
        }
        ast_expr::Kind::BinOp(op, lhs, rhs) => visit_binop(ctx, op, lhs, rhs)?,
        ast_expr::Kind::Function(func) => {
            let mut ctx_ = Context::default();
            function::visit(&mut ctx_, func)?;
            ctx.add_child(ctx_);
        }
        ast_expr::Kind::Call(callee, args) => visit_call(ctx, callee, args)?,
        ast_expr::Kind::If(cond, then_, else_) => visit_if(ctx, cond, then_, else_)?,
        ast_expr::Kind::Return(val) => visit_ret(ctx, val)?,
        ast_expr::Kind::Exprs(exprs) => {
            for expr in exprs {
                visit(ctx, expr)?;
            }
        }
    }
    Ok(())
}

fn visit_binop(
    ctx: &mut Context,
    op: &ast_expr::BinOpKind,
    lhs: &ast_expr::Node,
    rhs: &ast_expr::Node,
) -> Result<()> {
    visit(ctx, lhs)?;
    visit(ctx, rhs)?;
    match op {
        ast_expr::BinOpKind::Sub => ctx.push(Inst::Sub),
        ast_expr::BinOpKind::Mul => ctx.push(Inst::Mul),
        ast_expr::BinOpKind::Eq => ctx.push(Inst::Eq),
        _ => todo!(),
    }
    Ok(())
}

fn visit_call(ctx: &mut Context, callee: &ast_expr::Node, args: &[ast_expr::Node]) -> Result<()> {
    for arg in args {
        visit(ctx, arg)?;
    }
    match callee.kind() {
        ast_expr::Kind::Ident(name) => {
            ctx.push(Inst::PushStr(name.to_owned()));
        }
        _ => todo!(),
    }
    ctx.push(Inst::Call);
    Ok(())
}

fn visit_if(
    ctx: &mut Context,
    cond: &ast_expr::Node,
    then_: &ast_expr::Node,
    else_: &Option<Box<ast_expr::Node>>,
) -> Result<()> {
    assert!(else_.is_none());
    visit(ctx, cond)?;
    let cur = ctx.code.len() as i32;
    ctx.push(Inst::Jne(0));
    visit(ctx, then_)?;
    let merge = ctx.code.len() as i32;
    match ctx.code.get_mut(cur as usize).unwrap() {
        Inst::Jne(ref mut offset) => {
            *offset = merge - cur;
        }
        _ => panic!(),
    }
    Ok(())
}

fn visit_ret(ctx: &mut Context, val: &ast_expr::Node) -> Result<()> {
    visit(ctx, val)?;
    ctx.push(Inst::Ret);
    Ok(())
}
