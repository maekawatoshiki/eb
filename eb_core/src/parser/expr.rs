use super::Context;
use crate::ast::expr;
use anyhow::Result;

pub fn parse(ctx: &mut Context) -> Result<expr::Node> {
    let loc = ctx.cur_loc()?;
    let ident = ctx
        .expect_any_ident()?
        .kind()
        .as_ident()
        .unwrap()
        .to_string();
    Ok(expr::Node::new(expr::Kind::Ident(ident), loc))
}
