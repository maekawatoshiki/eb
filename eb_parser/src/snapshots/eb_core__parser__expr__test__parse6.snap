---
source: src/parser/expr.rs
expression: "parse(&mut ctx).expect(\"fail to parse\")"

---
Node {
    kind: Call(
        Node {
            kind: Ident(
                "f",
            ),
            loc: Location(
                0,
            ),
        },
        [
            Node {
                kind: Int(
                    1,
                ),
                loc: Location(
                    2,
                ),
            },
            Node {
                kind: Ident(
                    "x",
                ),
                loc: Location(
                    5,
                ),
            },
        ],
    ),
    loc: Location(
        1,
    ),
}
